src/modules/ekf2/EKF/ekf_helper.cpp

```diff
@@ -85,156 +85,170 @@ bool Ekf::checkAltitudeValidity(const float altitude)
 }
 
-bool Ekf::setEkfGlobalOriginFromCurrentPos(const double latitude, const double longitude, const float altitude,
-		const float eph, const float epv)
+bool Ekf::resetGlobalPositionTo(const double latitude, const double longitude, const float altitude,
+				const float hpos_var, const float vpos_var)
 {
-	if (!setLatLonOriginFromCurrentPos(latitude, longitude, eph)) {
+	if (!resetLatLonTo(latitude, longitude, hpos_var)) {
 		return false;
 	}
 
 	// altitude is optional
-	setAltOriginFromCurrentPos(altitude, epv);
+	resetAltitudeTo(altitude, vpos_var);
 
 	return true;
 }
 
-bool Ekf::setLatLonOriginFromCurrentPos(const double latitude, const double longitude, const float eph)
+bool Ekf::resetLatLonTo(const double latitude, const double longitude, const float hpos_var)
 {
 	if (!checkLatLonValidity(latitude, longitude)) {
 		return false;
 	}
 
-	_pos_ref.initReference(latitude, longitude, _time_delayed_us);
+	Vector2f pos_prev;
+
+	if (!_local_origin_lat_lon.isInitialized()) {
+		MapProjection zero_ref;
+		zero_ref.initReference(0.0, 0.0);
+		pos_prev = zero_ref.project(_gpos.latitude_deg(), _gpos.longitude_deg());
+
+		_local_origin_lat_lon.initReference(latitude, longitude, _time_delayed_us);
+
+		// if we are already doing aiding, correct for the change in position since the EKF started navigating
+		if (isLocalHorizontalPositionValid()) {
+			double est_lat;
+			double est_lon;
+			_local_origin_lat_lon.reproject(-pos_prev(0), -pos_prev(1), est_lat, est_lon);
+			_local_origin_lat_lon.initReference(est_lat, est_lon, _time_delayed_us);
+		}
+
+		ECL_INFO("Origin set to lat=%.6f, lon=%.6f",
+			 _local_origin_lat_lon.getProjectionReferenceLat(), _local_origin_lat_lon.getProjectionReferenceLon());
 
-	// if we are already doing aiding, correct for the change in position since the EKF started navigating
-	if (isLocalHorizontalPositionValid()) {
-		double est_lat;
-		double est_lon;
-		_pos_ref.reproject(-_state.pos(0), -_state.pos(1), est_lat, est_lon);
-		_pos_ref.initReference(est_lat, est_lon, _time_delayed_us);
+	} else {
+		pos_prev = _local_origin_lat_lon.project(_gpos.latitude_deg(), _gpos.longitude_deg());
 	}
 
-	if (PX4_ISFINITE(eph) && (eph >= 0.f)) {
-		_gpos_origin_eph = eph;
+	_gpos.setLatLonDeg(latitude, longitude);
+	_output_predictor.resetLatLonTo(latitude, longitude);
+
+	const Vector2f delta_horz_pos = getLocalHorizontalPosition() - pos_prev;
+
+#if defined(CONFIG_EKF2_EXTERNAL_VISION)
+	_ev_pos_b_est.setBias(_ev_pos_b_est.getBias() - delta_horz_pos);
+#endif // CONFIG_EKF2_EXTERNAL_VISION
+
+	updateHorizontalPositionResetStatus(delta_horz_pos);
+
+	if (PX4_ISFINITE(hpos_var)) {
+		P.uncorrelateCovarianceSetVariance<2>(State::pos.idx, math::max(sq(0.01f), hpos_var));
 	}
 
+	// Reset the timout timer
+	_time_last_hor_pos_fuse = _time_delayed_us;
+
 	return true;
 }
 
-bool Ekf::setAltOriginFromCurrentPos(const float altitude, const float epv)
+bool Ekf::resetAltitudeTo(const float altitude, const float vpos_var)
```

Comment:
find better names than resetAltitudeTo and resetHeightTo; the difference is not directly clear